Massive enhancements

abhay-s-rawat opened the pull request:
- Massive upgrades without breaking existing functionality.
- 2 New functions extractCoverWithFuture and exportVideoWithFuture both functions will do same task as default export functions but they will behave like Future. Their return type is Future<File?> and they does not have onCompleted callback. To avoid nests of callback users who want to use future version they can use these functions as well.
Updated example on how to use them.
- New capDimension parameter in export functions in which user can limit the max dimension without losing video aspect ratio.

- Now user will have more control over output video
cappedVideoBitRate nullable parameter (pass 128000 for 128kbps)- will cap the max bitrate of output video
cappedAudioBitRate nullable parameter (pass 1500000 for 1.5Mbps) - will cap the max bitrate of output audio
cappedOutputVideoSize nullable parameter - Advanced feature when passed a size in bytes will limit the output file size.
- cappedOutputVideoSize working
    - case 1: If cappedOutputVideoSize is not passed _maxDuration parameter value will limit the trim.
    - case 2: cappedOutputVideoSize is passed , for example 16MB like whatsapp
      - Step 1: Now on-initialize ffprobe will get the metadata of video to fetch current bitrates.
      - Step 2: updateMaxDuration function will calculate the max duration a user can select which falls below the cappedOutputVideoSize size.
        - example 1: A low bitrate file of 16Mb can be of 2-3 minutes or more , trim slider will be adjusted to that max value.
        - example 2: A high bitrate file whose maxduration was sepcified to be 60 seconds.
          - case 1: cappedVideoBitRate and cappedAudioBitRate are present
            - trimslider will be adjusted according to the cappedVideoBitRate and cappedAudioBitRate.
          - case 2: cappedVideoBitRate and cappedAudioBitRate are absent
            - trimslider will be adjusted accroding to the maximum duration that can be attained which falls below cappedOutputVideoSize.

- Mute audio functionality given through a value notifier and ffmpeg command is adjusted according to that.

- Output size prediction value notifier is introduced to predict the output size through bitrates and trimmed duration . variable name estimatedOutputSize.
It can be used with Value listenable builder to show size like whatsapp does. Updated example that shows the filesize and the trimmed duration.

- Introduced Trimmed duration (trimmedDuration) value notitifer to show the duration as user uses trim. Updated example to show that duration.

- croppedDimensions variable will always have latest value which can be used by user to show or operate on. Initially those dimension will be of video size but then as video crop is applied it will have latest dimensions which will be applied on output.

- Added screenshot for new updated exmaple in assets/readme

abhay-s-rawat committed:
Massive updates

abhay-s-rawat committed:
Massive updates

abhay-s-rawat committed:
Massive updates

abhay-s-rawat committed:
Massive updates

abhay-s-rawat commented:
And some bug fixes done like on exporting :
// (minCrop >= _min && maxCrop <= _max) was returning true always which was triggering scale always
    if (((minCrop == _min) && (maxCrop == _max)) == false) {
      filters.add(await _getCrop());
    }

abhay-s-rawat commented:
I previously did those changes but I fall behind too much in commits so I freshly did those changes.
In earlier time I made the trim slider ui look like whatsapp, will do it sometime in later update.

abhay-s-rawat committed:
Massive updates

abhay-s-rawat committed:
Changelog update

abhay-s-rawat commented:
Hello , anybody there ?

abhay-s-rawat commented:
Any reviews maintainers ? I am currently using my github version in my project. If you can push this to pubdev I can use that before I more project to production.

LeGoffMael commented:
Thanks for your contribution

I am in holidays, I will review it later

abhay-s-rawat commented:
Ok please don't forget it. Thanks

LeGoffMael commented:
Hey!
I finally got time to look into your pull request.
Thanks a lot for your work but since there is so many changes it is a bit difficult to review
So i have some questions to understand well

- I wanted to ask why you added `extractCoverWithFuture` and `exportVideoWithFuture` functions ? Is there any issue with the current existing exporting functions ? for me it just looks like twice as more code to maintain.

- What is the purpose of `capDimension` parameter ? Isn't using `maxCrop` function doing the same thing ?

- What is the purpose of `dimensionDivisibleBy` ? would export a mp4 video with a size indivisible by 2 lead to an error ?
If so is this parameter needed ? it sounds to me that it should be ceiled by default, without the need of a parameter.

- `estimatedOutputSize` don't seems to be updated when cropping the video

LeGoffMael commented on lib/domain/bloc/controller.dart:
```diff
@@ -1,8 +1,11 @@
+import 'dart:async';
+import 'dart:developer';
```
in flutter 3.3.0 this import don't seems to be used

abhay-s-rawat replied:
Added just for logging.

LeGoffMael commented on lib/domain/bloc/controller.dart:
```diff
@@ -222,13 +297,105 @@ class VideoEditorController extends ChangeNotifier {
+  }
+
+  _setCurrentBitRates() async {
+    Completer completer = Completer();
```
This function return a `Future<void>`,is this `Completer` needed ?

abhay-s-rawat replied:
Yes it has a function that has a callback, which he has to wait for.

abhay-s-rawat commented:
> * I wanted to ask why you added `extractCoverWithFuture` and `exportVideoWithFuture` functions ? Is there any issue with the current existing exporting functions ? for me it just looks like twice as more code to maintain.
Ans => Yes the content which contains filter is common so I created seperate function to return execute command. These functions are added to avoid nested callbacks problem (oftenly known as Callback hell among developers) which I ran into. So I added Future versions of same functions.

> * What is the purpose of `capDimension` parameter ? Isn't using `maxCrop` function doing the same thing ?
Ans => maxCrop will limit the amount you can crop. capDimension will cap the exported dimension. ex: a 1920 x 1080 pixels 16:9 video will be pretty heavy but you can cap its dimemsions . capDimension = 640 means its largest side will be capped to 640 and its other side will be reduced as per aspect ratio, It does not have effect if video's height and width are smaller than capDimension.
1920 x 1080 pixels with capDimension = 640 add a scale command to ffmpeg , something like "scale=640:-1" or "scale=-1:640".
This approach is best to tackle heavy images/videos. I observed whatsapp uses quite same approach when image is greater than particular size it caps its dimensions to 640 max.


> * What is the purpose of `dimensionDivisibleBy` ? would export a mp4 video with a size indivisible by 2 lead to an error ? >   If so is this parameter needed ? it sounds to me that it should be ceiled by default, without the need of a parameter.
Ans => See h264 needs dimensions after cropping to be a multiple of 2 (even number otherwise will throw error) and h265 needs side to be a multiple of 8.  This happens when user wants to switch encoders. h265 has better quality in same bitrates as h264 but it utilizes more cpu. So I thought to add a option for same.


> * `estimatedOutputSize` don't seems to be updated when cropping the video
Ans = > Actually I added this functionality to add an estimate for output, it won't be accurate all the time but it will be near the output size.
Eg: assume output file size is capped to 16mb like whatsapp
You chose a video which has 4mbps video and 256kbps audio bitrates. 5min video for example.
Such high birates video will easily cross 16mb limit for few seconds, imagine you can only send few seconds of video from it.

This is where bitrate capping came into role it will enfore a constant bitrate of 1mbps+128kbps for example if set so. So if you crop such video files it's output size wont be affected because even if you crop 50 perecent section, it can still enforce the limit because it halfed the dimensions of video frame which is still heavier than a 1mbps(each frame is still heavier than our cap).

So a video with 25 percent dimensions will have better pixels quality than a 50 percent cropped video. Because 50 percent cropped video will have frame size reduced to half thats equal to frame size in 2mbps video then a CBR of 1mbps is applied which reduces quality to 50 percent.

I will try to make it more accurate in future but as of now its just for prediction and limiting the duration of trim when max size is enforced.

I will study ffmpeg more and add more functionality , last time I tried to add a instagram like add text to video according to duration but I got diverted towards my project so was not able to do so.

abhay-s-rawat commented:
Basically its crop with minimal to no quality loss.
I made the trim look like whastapp last time too , like its in ios. Next update will be to make it add text like instagram. If there is anything you want to suggest . Please suggest I am using this package in my project and I want it to be best (it already is best).

abhay-s-rawat commented:
Is everything working fine admin ?

abhay-s-rawat commented:
Any update admin ?

LeGoffMael commented:
I am busy on my main project, sorry but i will review it when i have some times to focus completely to this library.

abhay-s-rawat commented:
Sure, take your time. Thanks

abhay-s-rawat commented:
Did you got time to look into this ?

abhay-s-rawat commented:
Admin did you got time to look into this ? I am waiting since August.

LeGoffMael commented:
I will probably be able to work on this package next month. You should be able to use your fork branch until then.

LeGoffMael commented:
I am working on your PR, i cleaned it and remove the new export future functions (i think it should be discussed if really useful and there is enough change in this PR without it).

You can see it on this branch : https://github.com/seel-channel/video_editor/tree/capped-params

My only concerned, is that the TrimSlider was not updated after the maxDuration was updated (from mute/unmute button). I tried to make this feature on the above branch.

Please let me know if this branch answer apply your changes properly.

abhay-s-rawat commented:
Export future functions will not affect existing functionality but new users can use async/await rather than witing for callback execution. More nested the callbacks become ,it becomes more difficult to control the flow.

See currently if a user wants a cover and a video how he will write code .
Non future versions:
User will call cover export function and in its callback he will wait and then he calls video export function inside that callback and when that callback finishes he may want to do other things, thus creating callback nest.

Future version:
User will call cover export function he/she will await and then does same for video export.

LeGoffMael commented:
OK i got your point, i will add it while minimizing the code duplication

LeGoffMael commented:
> My only concerned, is that the TrimSlider was not updated after the maxDuration was updated (from mute/unmute button). I tried to make this feature on the above branch.

What are your thoughts regarding my second point ?

LeGoffMael commented:
the new future function has been added in the branch https://github.com/seel-channel/video_editor/tree/capped-params

If you can tell me your thoughts about the TrimSlider update, and confirm that this branch is working as intended for you i would be happy to merge it for the next version

abhay-s-rawat commented:
> > My only concerned, is that the TrimSlider was not updated after the maxDuration was updated (from mute/unmute button). I tried to make this feature on the above branch.
> 
> What are your thoughts regarding my second point ?

It was working on my copy. Will check again.

LeGoffMael commented:
I checked again your branch and even though the `trimmedDuration` is updated, the TrimSlider is not.

I tried to implement this behavior on the branch [capped-param](https://github.com/seel-channel/video_editor/tree/capped-params).
I recorded my screen to makes it easier to understand :

| In your fork | In [capped-param](https://github.com/seel-channel/video_editor/tree/capped-params) |
| -------------------------------- | --------------------------------- |
|  <video src="https://user-images.githubusercontent.com/22376981/199658603-2fb715ce-a33b-4eb7-a13c-1f01e5b3a8cb.MP4" /> |  <video src="https://user-images.githubusercontent.com/22376981/199658633-dbddb381-96bd-4a26-ac89-8397546ed12a.MP4" /> |

abhay-s-rawat commented:
I will look into this and will do another pull request but keep this open for tracking.

abhay-s-rawat commented:
FFmpeg kit is available for macOS also , can we add macos support to this package ?
And I was able to do ffmpeg on windows too, will that be ok if I add support for windows too ?

LeGoffMael commented:
Unfortunately [video_thumbnail](https://pub.dev/packages/video_thumbnail) library supports only iOS and android so I don't think we are ready to support new platforms yet

abhay-s-rawat commented:
ffmpeg can generate thumbnails, I can make it work, worry not .

LeGoffMael commented:
If the performances are good it would be great indeed!
but please submit this feature in a specific PR, thanks for your effort !

abhay-s-rawat commented:
Will post this in seperate PR. Most probably need to bump version to 2.0.0.

abhay-s-rawat commented:
Published ffmpeg_helper on dart.dev.
https://pub.dev/packages/ffmpeg_helper

It can easily construct commands and run ffprobe. It can generate thumbnails also on all platforms except WEB.

abhay-s-rawat commented:
Will update video editor later and bump to 2.0.0 in separate PR.

LeGoffMael commented:
> Published ffmpeg_helper on dart.dev. https://pub.dev/packages/ffmpeg_helper
> 
> It can easily construct commands and run ffprobe. It can generate thumbnails also on all platforms except WEB.

I don't think there is the need for another package, flutter_kit_ffmpeg can generate thumbnails already

abhay-s-rawat commented:
But it cannot run on windows. For android ios and mac I am using same
pacakage

On Mon, 26 Dec 2022 at 6:46 AM, Maël ***@***.***> wrote:

> Published ffmpeg_helper on dart.dev.
> https://pub.dev/packages/ffmpeg_helper
>
> It can easily construct commands and run ffprobe. It can generate
> thumbnails also on all platforms except WEB.
>
> I don't think there is the need for another package, flutter_kit_ffmpeg
> can generate thumbnails already
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/LeGoffMael/video_editor/pull/104#issuecomment-1364782633>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AUIPCBAWUEQQMQN3S7O7MALWPDWWDANCNFSM57EWQ4JA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

abhay-s-rawat commented:
It will work on linux too.I will test and update asap.

On Mon, 26 Dec 2022 at 11:09 AM, Abhay Rawat ***@***.***>
wrote:

> But it cannot run on windows. For android ios and mac I am using same
> pacakage
>
> On Mon, 26 Dec 2022 at 6:46 AM, Maël ***@***.***> wrote:
>
>> Published ffmpeg_helper on dart.dev.
>> https://pub.dev/packages/ffmpeg_helper
>>
>> It can easily construct commands and run ffprobe. It can generate
>> thumbnails also on all platforms except WEB.
>>
>> I don't think there is the need for another package, flutter_kit_ffmpeg
>> can generate thumbnails already
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/LeGoffMael/video_editor/pull/104#issuecomment-1364782633>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AUIPCBAWUEQQMQN3S7O7MALWPDWWDANCNFSM57EWQ4JA>
>> .
>> You are receiving this because you authored the thread.Message ID:
>> ***@***.***>
>>
>

abhay-s-rawat commented:
Now I made things work on linux too.

https://pub.dev/packages/ffmpeg_helper
v 1.0.0
